File: server/projects.js
import * as fs from "node:fs";
import * as path from "path";
import getConfigValue from "../config.js";

/**
 * Get the name of the parent directory for the projects
 */
const projectsDir = getConfigValue("projectsDir");

/**
 * Turn all projects into objects and get a list of them
 * @returns {Project[]} Array of Project objects
 */
export default function getProjects() {
  const contents = fs.readdirSync(projectsDir);
  const projects = [];

  for (let i = 0; i < contents.length; i++) {
    if (fs.existsSync(path.join(projectsDir, contents[i], "package.json"))) {
      projects.push(new Project(contents[i]));
    }
  }

  return projects;
}

/**
 * Represents a project in the projects directory
 * @constructor
 * @typedef {Object} Project
 * @param {string[]} project
 * @prop {string} name        - The identifier of the project
 * @prop {string} root        - The root of the project
 * @prop {string} fancyName   - The name of the project
 * @prop {string} version     - The current version of the project
 * @prop {string} description - A short description of the project
 * @prop {string} entry       - The entry point of the project
 */
function Project(project) {
  this.name = project;
  this.root = path.join(projectsDir, project);

  const packageJson = JSON.parse(fs.readFileSync(path.join(this.root, "package.json")));

  this.fancyName = packageJson.name;
  this.version = packageJson.version;
  this.description = packageJson.description;
  this.entry = path.join(this.root, "index.html");
}

getProjects();

File: server/projects.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path"
)

type Project struct {
	Name        string
	Root        string
	Entry       string
	FancyName   string `json:"name"`
	Version     string `json:"version"`
	Description string `json:"description"`
}

const projectsDir string = "./projects"

func getProjects() []Project {
	contents, err := os.ReadDir(projectsDir)
	if err != nil {
		log.Fatal(err)
	}

	var projects []Project
	for _, position := range contents {
		if !position.IsDir() {
			continue
		}

		packageJson := path.Join(projectsDir, position.Name(), "package.json")
		if _, err := os.Stat(packageJson); err != nil {
			continue
		}

		projects = append(projects, *projectConstructor(position.Name(), packageJson))

	}
	return projects
}

func projectConstructor(project string, packageJson string) *Project {
	p := new(Project)
	data, _ := os.ReadFile(packageJson)
	if err := json.Unmarshal(data, &p); err != nil {
		fmt.Println("Error:", err)
	}
	p.Name = project
	p.Root = path.Join(projectsDir, project)
	p.Entry = path.Join(p.Root, "index.html")

	return p
}
